### PR TITLE
Decrease parallelism in CI tests

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR decreases the parallelism in the CI tests by overriding the environment variable `MAVEN_PARALLELISM` when running the unit and integration tests. It reduces it to 2 for both, which means we'll run tests for two modules in parallel, and each module will run `$MAVEN_PARALLELISM / 2` forks, which totals one fork per CPU. It also enforces the number of forks to 6 via the environment variable `SUREFIRE_FORKCOUNT`. It also introduces the possibility of overriding the fork count via an environment variable so we can play around with different values without making new changes.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
Blocked by https://github.com/zeebe-io/zeebe/pull/5929 and https://github.com/zeebe-io/zeebe/pull/5932

File: .ci/scripts/distribution/it-java.sh
Comment (korthout) [inline .ci/scripts/distribution/it-java.sh:4]: Perhaps we should just default this explicitly to 1

File: Jenkinsfile
Comment (korthout) [inline Jenkinsfile:219]: Let's also set `FORK_COUNT = 2 ` or so, currently this PR seems to slow down the IT by a lot.

**Review by korthout — COMMENTED** [review]
Good changes! This is gonna improve things a lot over time. Just some small questions/remarks

File: .ci/scripts/distribution/it-java.sh
Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:3]: `$(getconf _NPROCESSORS_ONLN)` returns the number of processors available in a POSIX compliant way. If `LIMITS_CPU` isn't set, then this is the value we assign. I'm happy to omit it, which keeps the behaviour as before - it breaks if undefined :)

File: .ci/scripts/distribution/it-java.sh
Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:4]: That's different behaviour though. Right now, this is set up so you can override whatever's defined in the POM - if we explicitly set to 1, then we __always__ override it, which I don't know if we really want to do. I think ignoring it if unset is a little more intuitive in my opinion.

File: Jenkinsfile
Comment (npepinpe) [reply Jenkinsfile:219]: I'm confused how it could affect it at all - there's only two modules, so lowering the limits here shouldn't matter - it can still only process two modules in parallel :thinking:

Comment (npepinpe) [reply Jenkinsfile:219]: It takes 20 minutes to process the IT stage because the update tests take 20 minutes. This is independent of this PR :slightly_smiling_face:

File: .ci/scripts/distribution/it-java.sh
Comment (korthout) [reply .ci/scripts/distribution/it-java.sh:3]: Just leave it as is, I'm just unfamiliar with this config property.

File: Jenkinsfile
Comment (korthout) [reply Jenkinsfile:219]: Ah, I missed that. But that's totally fine then. I didn't consider that the default fork count from the pom would be used. Let's leave it like this.

**Review by korthout — APPROVED** [review]
LGTM 🚀

File: .ci/scripts/distribution/it-java.sh
Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:3]: Yeah it looks weird, but it's the only POSIX way I know of getting the cpu count (which also works on MacOS ;))

**Comment by npepinpe** [comment]
bors cancel

**Comment by npepinpe** [comment]
I realized why we set the `MaxRAMFraction` :sweat_smile: As we're creating many forks, each of these could (but most likely won't) theoretically use up all the memory. So, when you don't set anything, the MaxHeapSize defaults to 25% of your memory. When we fork, it uses a copy-on-write mechanism such that initially the memory isn't much bigger, but as you write it will start copying pages and using more memory. So if we have many many forks, and they all think they can use up to 25% of the memory, you run into issues such as `script exiting with error code 137` which indicate the oom-killer just killed your process.

So, let's think about it a bit. We now limit it to 2 modules in parallel, each with 6 forks. So we have 12 processes which could (theoretically, but most likely won't) use 25% of the memory each :slightly_smiling_face: We have two options, one is setting the `MaxRAMFraction` to 12 - however we don't really know that from the script (this is inferred from the `pom.xml`). So, not sure what's the best solution here, would be nice to quickly zoom and brainstorm about it.

**Comment by npepinpe** [comment]
I also fixed some merge conflicts after templating was introduced for the pod spec.

**Review by korthout — APPROVED** [review]
LGTM 🎸

**Comment by npepinpe** [comment]
So turns out we never configured failsafe to run in parallel - which means all tests ending in `IT` (and not `Test`) were not using the `forkNumber` - so when assigning ports, they were all using the same port range :exploding_head:

**Review by korthout — APPROVED** [review]
Haha wow that's a good find! This appears more robust. LGTM 🍰

**Comment by korthout** [comment]
@npepinpe maybe first squash before merging

**Comment by npepinpe** [comment]
> @npepinpe maybe first squash before merging

:+1: Let's let it run at least once before merging too :slightly_smiling_face:

**Comment by npepinpe** [comment]
bors r+

It failed due to a channel close exception, however I think we can give it a go.

**Comment by korthout** [comment]
@npepinpe maybe just restart the build for the branch 10 times to see if this is more / less flaky then before

**Comment by npepinpe** [comment]
Well it's not more or less flaky than develop so far if you look at develop :sweat_smile: 

I asked infra about the weird error, i.e. `>> ./target/FlakyTests.txt` failing because the target file doesn't exist (when `>>` should be creating it if it doesn't??).

**Comment by npepinpe** [comment]
So Ben had a good idea, which is possibly `target` may not exist - I'm actually unsure why we write there in the first place anyway. It actually doesn't matter too much now, as I discovered (and will open another PR) with Christian that currently the post steps run in the default agent, which means we are not gathering flaky test nor test coverage info from IT.

**Comment by npepinpe** [comment]
bors retry

File: .ci/scripts/distribution/it-java.sh
Comment (korthout) [inline .ci/scripts/distribution/it-java.sh:32]: @npepinpe don't forget this change then for the other scripts.

Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:32]: I think on the main agent it doesn't fail because we build the root project, whereas on the IT we only build specific modules, but I can gladly change it.

Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:32]: You can confirm this by removing your `./target` folder from your build and running verify for specific modules, it won't build it. But if you run for the project it will appear. Why it was however sometimes appearing for IT as well is a mystery to me.

**Comment by npepinpe** [comment]
This time, in a fun twist, it failed to fetch some docker images (e.g. `camunda/zeebe:0.25.0`) :smile: What error will be next? Place your bets! :drum:

**Comment by npepinpe** [comment]
Now the build was aborted - this is something Christian explained to me. Every auto scaling event can cause builds (across the whole CI pool, not just Zeebe's) to get aborted. So the more builds are ran, the more likely this is to happen. The more agents we have, the more likely - the longer each agent runs, the more likely, etc.

So starting 10 builds will probably cause some to be aborted (and other's branches too :slightly_smiling_face:)

**Comment by npepinpe** [comment]
And again it fails to fetch docker images. I'm not sure my PR has anything to do with this, but I'm happy to look into it because, at this point, it's not like it can get worst.

That said, I would point out that if on every change we make to CI, and something that seems unrelated happens, we end up spending days because we can't make a decision on whether or not it's related - clearly this is a bad state to be in, and shows that we really don't know what we're doing (or we're working on a very shaky foundation). Either way, something that needs to change.

**Comment by npepinpe** [comment]
@korthout imo I don't think the current flakiness with these builds has to do with the changes in this PR, but I'd like to hear from your side :+1:

**Comment by korthout** [comment]
@npepinpe I agree, but I'd rather wait until tomorrow to retry some builds then and see that this really doesn't make things worse. Further investigation is not needed at this time IMO.

**Comment by npepinpe** [comment]
Running it 8 times in a row:

1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Pass
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Failed - Verify / IT / Test / BrokerAdminServiceTest(it-testrun).shouldResumeExporterAfterRestart (Flaky Test) : address in use
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate & RollingUpdateTest(it-testrun).shouldReplicateSnapshotAcrossVersions (Flaky Test) : failed to pull docker image
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Passed
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image **and** TEST-io.zeebe.broker.it.clustering.SnapshotReplicationTest-it-testrun.xml.[failed-to-read]

Last 8 builds on develop:

1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Failed - TEST-io.zeebe.broker.it.clustering.SnapshotReplicationTest-it-testrun.xml.[failed-to-read] **and** Verify / IT / Test / UpdateTest(it-testrun).updateWithSnapshot(String, UpdateTestCase)[7] (Flaky Test) (an assertion failed)
1. Failed - script returned exit code 137, which usually means OOM killed
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) (and 3 others in the same module) : failed to pull docker image
1. Pass
1. Failed - Verify / IT / Test / RollingUpdateTest(it-testrun).shouldPerformRollingUpdate (Flaky Test) : failed to pull docker image
1. Aborted due to connection error

**Comment by npepinpe** [comment]
Tailing the docker logs to figure out why we can't pull, one interesting thing is we do have collisions with ports used in our tests:

```
WARN[2020-12-02T16:12:17.256962414Z] Failed to allocate and map port: Error starting userland proxy: listen tcp 0.0.0.0:32822: bind: address already in use, retry: 1
WARN[2020-12-02T16:12:29.791267840Z] Failed to allocate and map port: Error starting userland proxy: listen tcp 0.0.0.0:32830: bind: address already in use, retry: 1 
WARN[2020-12-02T16:12:29.852227338Z] Failed to allocate and map port: Error starting userland proxy: listen tcp 0.0.0.0:32834: bind: address already in use, retry: 1 
```

**Comment by korthout** [comment]
>we do have collisions with ports

do you mean in this PR or on develop?

**Comment by npepinpe** [comment]
On develop.

**Comment by npepinpe** [comment]
Regarding the docker images, I'll look into it with Christian tomorrow. I would merge this PR as it is, since as I showed, it's not flakier than develop. I'm happy to wait until tomorrow, and all the new and exciting issues we'll discover :smile:

**Review by korthout — APPROVED** [review]
Those results looks much more promising now. ✅

**Pull request merged**
